### ISSUE TITLE
MSSQL Comments in SchemaBuilder are not used, using exec sp_updateextendedproperty in lieu of comments

### ISSUE THREAD

**Issue by alexanderankin** [knex/knex#4204]
# Environment

Knex version: 21.12
Database + version: mssql (just the dialect)
OS: ubuntu

Select applicable template from below.
If issue is about MSSql tag @smorey2.

# Feature discussion / request

1. Explain what is your use case

I would like to use the `db.schema.createTable('mytable', t => t.string('somecol').comment('the col'));` method and have the resulting comment be persisted in the database. The warning is presented to the user if the comment is too big (instead of giving feedback that its not supported) so lets support it.

2. Explain what kind of feature would support this

Here's background on what this is: [web.archive.org http://sqlblog.com/blogs/michael.../2010/01/12/t-sql-tuesday-easy-extended-properties.aspx](https://web.archive.org/web/20170716134332/http://sqlblog.com/blogs/michael_coles/archive/2010/01/12/t-sql-tuesday-easy-extended-properties.aspx). basically the suggested proposal below implements this, which is how you can add comments using the visual tools provided with [SSMS](https://docs.microsoft.com/en-us/sql/ssms/sql-server-management-studio-ssms?view=sql-server-ver15).

3. Give some API proposal, how the feature should work

I would like to add a PR with warnings where the column information is discarded.

Ideally knex would implement the real functionality (not just discarding values) where possible, here's a sneak preview of changes I would like to submit:

<details>
<summary>the diff</summary>

```diff
diff --git a/lib/dialects/mssql/schema/columncompiler.js b/lib/dialects/mssql/schema/columncompiler.js
index c8563b62..90925f37 100644
--- a/lib/dialects/mssql/schema/columncompiler.js
+++ b/lib/dialects/mssql/schema/columncompiler.js
@@ -93,11 +93,14 @@ Object.assign(ColumnCompiler_MSSQL.prototype, {
   comment(comment) {
     if (comment && comment.length > 255) {
       this.client.logger.warn(
         'Your comment is longer than the max comment length for MSSQL'
       );
     }
+
+    this.tableCompiler.addExtendedPropertyToColumn(comment, this.args[0]);
+
     return '';
   },
 });
 
 module.exports = ColumnCompiler_MSSQL;
diff --git a/lib/dialects/mssql/schema/tablecompiler.js b/lib/dialects/mssql/schema/tablecompiler.js
index 6ed8d6af..e7e5f877 100644
--- a/lib/dialects/mssql/schema/tablecompiler.js
+++ b/lib/dialects/mssql/schema/tablecompiler.js
@@ -97,12 +97,42 @@ Object.assign(TableCompiler_MSSQL.prototype, {
     );
   },
 
   // Compiles the comment on the table.
   comment() {},
 
+  addExtendedPropertyQuery: `EXEC sp_updateextendedproperty
+@name = N'MS_Description', @value = ?,
+@level0type = N'Schema', @level0name = ?,
+@level1type = N'Table',  @level1name = ?,
+@level2type = N'Column', @level2name = ?`,
+
+  addExtendedPropertyToColumn(comment, column) {
+    this.sequence.unshift({
+        sql: this.addExtendedPropertyQuery,
+        bindings: [
+          comment,
+          this.schemaNameRaw || 'dbo',
+          this.tableNameRaw,
+          column,
+        ]
+      });
+  },
+
+  addExtendedPropertyToTable(comment) {
+    this.sequence.unshift({
+        sql: this.addExtendedPropertyQuery,
+        bindings: [
+          comment,
+          this.schemaNameRaw || 'dbo',
+          this.tableNameRaw,
+          null,
+        ]
+      });
+  },
+
   changeType() {},
 
   // Renames a column on the table.
   renameColumn(from, to) {
     this.pushQuery(
       `exec sp_rename ${this.formatter.parameter(
```

</details>

and their example use case:

```sql
$ node -p "require('knex')({ client: 'mssql' }).schema.withSchema('sch').createTable('t', t => t.string('thing').comment('abc')).toString()" 
EXEC sp_updateextendedproperty
@name = N'MS_Description', @value = 'abc',
@level0type = N'Schema', @level0name = 'sch',
@level1type = N'Table',  @level1name = 't',
@level2type = N'Column', @level2name = 'thing';
CREATE TABLE [sch].[t] ([thing] nvarchar(255))
```

I am not super familiar with knex internals, and I've never needed to learn the internals of this one (unlike some other node libraries *tsk *tsk*) but in studying it more i found this detail. So, perhaps there is a much better way of implementing it. This is just the idea. thanks knex maintainers and contributors, for hearing me out.

**Comment by kibertoad** [comment]
@alexanderankin This looks very MSSQL-specific, so I would be wary of adding this as something that user has to be aware of and explicitly invoke, unless you can find equivalent cases for other databases - e. g. PostgreSQL has no such limits, as far as I'm aware, so you can just use comments directly. I'm open to adding this kind of a feature if it can be handled transparently inside of MSSQL dialect. If that is not possible, probably this should be left to a `knex.raw` operation.

**Comment by alexanderankin** [comment]
@kibertoad yep, by expanding, you can see in the diff references mssql dialect files. I'm proposing to reuse that same api, and add warnings to tell users when their data is discarded on the mssql dialect

**Comment by alexanderankin** [comment]
before, this `comment(comment)` method did nothing, so i added the `pushQuery` to dialect's table compiler which can add the comments to the internal metadata table, which is the mechanism for comments in that database (which i unfortunately have to use for work)

```diff
diff --git a/lib/dialects/mssql/schema/columncompiler.js b/lib/dialects/mssql/schema/columncompiler.js
index c8563b62..90925f37 100644
--- a/lib/dialects/mssql/schema/columncompiler.js
+++ b/lib/dialects/mssql/schema/columncompiler.js
@@ -93,11 +93,14 @@ Object.assign(ColumnCompiler_MSSQL.prototype, {
   comment(comment) {
     if (comment && comment.length > 255) {
       this.client.logger.warn(
         'Your comment is longer than the max comment length for MSSQL'
       );
     }
+
+    this.tableCompiler.addExtendedPropertyToColumn(comment, this.args[0]);
+
     return '';
   },
 });
```

**Comment by kibertoad** [comment]
@alexanderankin Shouldn't this code be inside the length check, so that it wouldn't be used for shorter comments?
This can be included, I guess, but definitely needs an integration test. See test/integration2/schema for examples.

**Comment by kibertoad** [comment]
Also needs a comment explaining what it is and why, and preferably a link to that article.

**Comment by alexanderankin** [comment]
@kibertoad yes it should probably only be used if its within the length (but maybe failing fast is good too?), I just wanted to include a minimal example for explanation. Is that the right way to get the table and schema names? I'm sure that there should be a better place to hardcode that 'dbo' (maybe also get a default from an option configurable in the client?), and for some reason, its generating the statements out of order.

I'm sure that I would be able to figure this out once i started working on a PR but I didnt want to spend time on something immediately obviously out of scope.

**Comment by kibertoad** [comment]
It seems generally correct, yes.

**Comment by jeremy-w** [comment]
i recently noticed my comments weren't making it from my schema to the db as well, so i'm happy to find this discussion!

is this still something you're working on @alexanderankin? (if not, i might have a go.)

i think we can also eliminate that 255-char warning. afaict from [the sp_addextendedproperty docs](https://docs.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-addextendedproperty-transact-sql?view=sql-server-ver15), the comment value limit is 7,500 bytes (!). it seems querying clients might need some config to request that many bytes judging by [this Q&A about descriptions coming back truncated at 256 chars](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/724f9edf-040f-4d00-8ec8-6d3d33d7b73c/truncated-msdescription-only-displays-first-256-characters?forum=transactsql), but OP reported the data was still present and accessible, and that after reconfiguring (or by explicitly casting before reconfiguration), they were able to access the full comment text, so we should be good to go for quite lengthy comments.

**Comment by alexanderankin** [comment]
sorry, this dropped off of my radar - please feel free to attempt to
implement this.

**Comment by kibertoad** [comment]
Released in 0.95.11
